### ISSUE TITLE
Enhancement: Detect/Warn on multilevel cycles

### ISSUE THREAD

**Issue by Olipro** [ayecue/greybel-vs#120]
This doesn't affect Greybel at all, but it can completely hang a script in-game. It requires a certain amount of cyclic references - simply doing `Foo = {}; Foo.__isa = Foo` doesn't affect it at all. This is a bug that's been fixed in the first 2023 release of MiniScript, but of course, GreyHack isn't using that.

From in-game testing, the hang seems to occur somewhere inside hash calculation of map keys - I suspect it's somewhere on the order of `O(n^2)` (or worse) since paring down the example below will hang the game for just a few seconds.

For the purposes of the enhancement, just warning on the detection of _any_ cycles would be a fantastic feature.

The bad code:

```Lua
Foo = {}
Foo.a = "str"
Foo.b = false
Foo.c = true
Foo.d = false
Foo.id = 0

Foo[0] = new Foo

Foo[1] = new Foo
Foo[1].id = 1
Foo[1].type = "str2"
Foo[1].en_str = "str3"
Foo.x = Foo[1]

Foo[2] = new Foo
Foo[2].id = 2
Foo[2].type = "str4"
Foo[2].en_str = "str5"
Foo.y = Foo[2]

Foo[3] = new Foo
Foo.z = Foo[3]

Foo[4] = new Foo
Foo[4].id = 2
Foo[4].type = "str6"
Foo[4].en_str = "str7"
Foo.a = Foo[4]

Foo[5] = new Foo
Foo[5].id = 2
Foo[5].type = "str6"
Foo[5].en_str = "str7"
Foo.b = Foo[5]

Bar = {}
Bar[new Foo[4]] = 3 //DEAD
```

**Comment by ayecue** [comment]
So your suggestion is to add some kind of warning message when running code with Greybel that has that cyclic pattern?

**Comment by Olipro** [comment]
Running, building, even while editing with a squiggly yellow underline if that's possible :)

Just something that mentions that if you're coding for GreyHack, you have a chance to produce a hanging, unkillable script.

**Comment by ayecue** [comment]
Well detecting this while building or editing will be pretty hard since I imagine that this requires some quite complex logic. Especially since there is no runtime logic available.

But detecting cyclic data while running the interpreter is a lot more feasible.

**Comment by Olipro** [comment]
Yeah, I suppose you could do a rudimentary check without executing by evaluating the script in the background - but the caveat there is that once any sort of map mutation is occurring in a `funcRef`, you have to skip over it since the script shouldn't actually execute any operations.

In any case, even just during runtime could prevent folks from shooting themselves in the foot.
